src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java

```diff
@@ -20,6 +21,53 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
         this.random = new SecureRandom();
     }
 
+    public Collection<Set<Hash>> getConnectedComponents(Set<Hash> transactions) throws Exception {
+        Set<Hash> unvisited = new HashSet<>(transactions);
+        List<Set<Hash>> result = new ArrayList<>();
+
+        // Outer loop iterates once for each component
+        while (!unvisited.isEmpty()) {
+            Set<Hash> currentComponent = new HashSet<>();
+
+            // Perform DFS scan to find all elements in this component
+            Deque<Hash> stack = new ArrayDeque<>();
+            Hash currentHash = unvisited.iterator().next();
+            stack.push(currentHash);
+            while (!stack.isEmpty()) {
+                currentHash = stack.pop();
+
+                if (!currentComponent.contains(currentHash)) {
+                    unvisited.remove(currentHash);
+                    currentComponent.add(currentHash);
+
+                    Collection<Hash> neighbors = getNeighbors(currentHash);
+                    Collection<Hash> unvisitedNeighbors = neighbors.stream()
```

Comment:
you don't actually need the set. so you can avoid collecting it and just pushing to the stack directly. using `forEach`.